documentation.md

```diff
@@ -0,0 +1,521 @@
+
+<figure class="image">
+  <img src="https://i.imgur.com/ri0Ej2o.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Simulated memory controller overview</figcaption>
+</figure>
+
+#### Top-level flow
+
+The top-level flow happens as follows:
+
+0. The response banks constantly advertise the identifier that will be allocated to the next address request.
+1. When the address request comes in:
+   - The delay calculator and the response banks are notified when an address request happens.
+   - The delay calculator stores metadata for its delay computation.
+   - The response banks reserve space for the corresponding responses that will subsequently come back from the real memory controller.
+2. When the corresponding responses comes back from the real memory controller, they are stored by the message banks.
+3. When the delay enables the release of some response and all the previous responses of the same AXI are released, the response banks transmit the response to the requester.
+4. When all the responses corresponding to a given address request have been transmitted to the requester, the response banks free the allocated memory for reuse.
+5.
+
+<figure class="image">
+  <img src="https://i.imgur.com/pbbpy0Y.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Simulated memory controller top-level flow</figcaption>
+</figure>
+
+## Response banks
+
+### High-level design
+
+The _simmem_resp_banks_ module is made of two distinct _simmem_resp_bank_ modules:
+
+- The write response bank, responsible for storing write responses.
+- The read data bank, responsible for storing read data. As read data can come back as burst data, this bank has to additionally support bursts, which is the essential difference with the write response bank.
+
+The two banks act independently.
+
+Each response bank has a FIFO interface with reservations, and stores messages in RAMs. FIFOs are implemented as linked lists to make best use of the RAM space.
+
+<figure class="image">
+  <img src="https://i.imgur.com/BpU4aZz.png" alt="Simulated memory controller internal overview">
+  <figcaption>Fig: Response bank top-level interface</figcaption>
+</figure>
+
+### Reservations
+
+The reservation mechanism built in the response banks provides simultaneously two features:
+
+1. It ensures that the responses to an address request will have space to be stored as soon as they are produced. This prevents inflated delays between request address acceptation and its response in the case of congestion in the response banks. This also prevents any deadlock situation.
+2. It provides internal identifiers, as soon as the address requests are made. This allows the delay calculator to immediately label the request metadata, for immediate processing. As the real memory controller may require multiple cycles to provide a response to an address request, producing an internal identifier as early as possible is a necessary feature.
```

Comment:
At this point it is not yet clear what metadata is. It has not yet been mentioned.

Reply:
The second sentence "As the real memory controller may require multiple cycles to provide a response to an address request, producing an internal identifier as early as possible is a necessary feature." is not clear. Why does a latency of potentially multiple cycles require a timely internal identifier?

Reply:
> At this point it is not yet clear what metadata is. It has not yet been mentioned.
Added precisely what the metadata is in this context, thanks!

Reply:
> The second sentence "As the real memory controller may require multiple cycles to provide a response to an address request, producing an internal identifier as early as possible is a necessary feature." is not clear. Why does a latency of potentially multiple cycles require a timely internal identifier?

The emphasis was not on the right place of the sentence. I rephrased to say that the iid has to be produced _as early as possible_.